Add events to Datadog and change metric name

wouterdebie opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
- Metric names are now `prefix.entity_id.state_attribute` (e.g. `hass.sun.rising` instead of `hass.sensor` with entity as a tag)
- entity_id, domain and state_attribute are now tags 

This means that you'll have to recreate dashboards and the like.

## Proposed change
<!--
Describe the big picture of your changes here to communicate to the   maintainers why we should accept this pull request. If it fixes a bug  or resolves a feature request, be sure to link to that issue in the additional information section.
-->
This PR adds events to Datadog for non-numerical state changes. Before, changes like "weather: cloudy" would be ignored, since they can't be expressed as a number. But now anything that isn't a number is captured as a datadog event.

Additionally this PR changes Datadog metric names to include the entity_id in order to correctly use metrics in datadog.
Before this commit metrics were named based on their domain, like `hass.sensor` or `hass.switch` and `entity_id`s were only expressed as tags. This however is somewhat incorrect, since in datadog a metic can have a type (gauge, rate or count) and a unit (like bits/second or ampere). Having all switches or all sensors express the same metric and distinguish makes it impossible to define a unit and type.

This commit refactors the datadog integration into the following:
- Metric names are now `prefix.entity_id.state_attribute` (e.g. `hass.sun.rising` instead of `hass.sensor` with entity as a tag)
- entity_id, domain and state_attribute are now tags 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Adds events to Datadog component

This commit adds events to Datadog for non-numerical state changes.
Before, changes like "weather: cloudy" would be ignored, since they
can't be expressed as a number. But now anything that isn't a number is
captured as a datadog event.

homeassistant commented:
Hi @wouterdebie, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#57293).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Changes Datadog metric names to include the entity_id

This commit changes Datadog metric names to include the entity_id in
order to correctly use metrics in datadog.
Before this commit metrics were named based on their domain, like
"hass.sensor" or "hass.switch" and entity_ids were only expressed as
tags. This however is somewhat incorrect, since in datadog a metic can
have a type (gauge, rate or count) and a unit (like bits/second or
ampere). Having all switches or all sensors express the same metric and
distinguish makes it impossible to define a unit and type.

This commit refactors the datadog integration into the following:
- Metric names are now "prefix.entity_id.state_attribute"
- entity_id, domain and state_attribute are now tags
